# Aula 12 - APIs

## Nossa agenda

1.  Como são formadas as APIs REST
1.  Como funciona o protocolo HTTP
1.  Como funciona o pacote Requests
1.  APIs com autenticação
1.  Escrita e leitura de arquivos em Python
1.  O que são arquivos CSV

## Me manda uma mensagem


#wilsontayar no [slack](https://terahq.slack.com)

/wilsontayar no [github](https://github.com/wilsontayar)

@wilsontayar no [twitter](https://twitter.com/wilsontayar)

----

## **1. Como são formadas as APIs REST**

Podemos dizer que **uma API é como se fosse uma página da web só que para computadores**. Só eles consegurem acessar, ler e interagir da maneira adequada com as APIs.

Como Data Scientists, vocês utilizarão APIs tanto internas (fornecidas pelas próprias empresas em que vocês trabalham) quanto externas. 

APIs são uma das principais formas que um Data Scientist utiliza para obter os dados que devem ser analisados.

---

Como exemplo, ao acessarmos a API da Jsonplaceholder (https://jsonplaceholder.typicode.com/posts/1), vemos o retorno abaixo, ao invés de uma página comum:

```json
{
  "userId": 1,
  "id": 1,
  "title": "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
  "body": "quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto"
}
```

Esse padrão de resposta é o que chamamos de **JSON**, e este é o **tipo de resposta mais comum entre as APIs**.

Ele foi feito para que máquinas entendam seus dados e os utilizem como quiser.

---


## Como funciona o protocolo HTTP

A internet trafega através do protocolo HTTP, ou **Hypertext Transfer Protocol**. O protocolo HTTP foi criado por Tim Berners-Lee, o criador da web, enquanto ele trabalhava na CERN em 1989 (sua equipe e ele também foram responsáveis pelo HTML).

O HTTP possui **duas entidades básicas**: a **Request** (pedido) e a **Response** (resposta)

Sempre que entramos em uma página, **enviamos uma request para o servidor** do site e **esperamos uma response** com o HTML da página.

Sempre que enviamos uma request devemos especificar um **método**.
A tabela abaixo cita os métodos mais comuns e explica brevemente o objetivo de cada um:

Método     | Objetivo
-----------|-----------
**GET**    | Obter informações
**POST**   | Salvar informações
**PUT**    | Atualizar informações
**DELETE** | Deletar informações

Todos eles são comumente utilizados por APIs. 

O método **GET** é o que seu navegador usa para baixar o HTML da página.

O metódo **POST** também é muito utilizado sempre que você envia algum formulário em um site.

**Além do método, as requests também possuem um header**

O **Header** (cabeçalho) tem a função de dizer para o servidor diversos dados sobre quem fez a solicitação.
Dessa forma, o servidor consegue até permitir que você veja mais informações dependendo do seu nível de acesso.


---

**Todos as respostas possuem, além do conteúdo da resposta, um código de status**.

Os códigos de status (status code) seguem um padrão.

Os mais comuns são:

Código  |  Significado
--------|---------------
200     | Ok
3XX     | O servidor vai te redirecionar para outra URL
4XX     | O recurso não foi encontrado, não está mais disponível, etc
5XX     | Algum problema aconteceu no servidor e ele não conseguiu processar a sua request



---

Outra característica das APIs REST é a **divisão dos recursos**. Por exemplo:

https://jsonplaceholder.typicode.com/users/1

Ao abrirmos esta URL estamos fazendo um **GET** para obtermos informações do usuário (/users) com código 1 (/1).
Podemos trocar o código do usuário para buscar outros usuários.

Também podemos pedir todos os usuários digitando:

https://jsonplaceholder.typicode.com/users

---

Também é muito comum o uso de parâmetros na URL para passarmos filtros para a API. Por exemplo:

https://jsonplaceholder.typicode.com/comments?postId=1

Perceba o "?postId=1" ao final da URL.

Sempre que precisamos passar parâmetros, começamos com um ponto de interrogação (?) e depois utilizamos o & para parâmetros futuros. Por exemplo:

https://jsonplaceholder.typicode.com/comments?postId=1&id=3

## Como funciona o pacote Requests

Como vimos anteriormente, o pacote requests do Python é utilizado para nos comunicarmos com URLs.

Podemos realizar GET, POST, PUT, DELETE e quaisquer outros métodos HTTP com ele.

Veja no exemplo abaixo como podemos fazer isso

In [1]:
import requests

api = requests.get('https://jsonplaceholder.typicode.com/users')

json = api.json()

for user in json:
    print('{0} ({1}) - {2}'.format(user['name'], user['email'], user['website']))

Leanne Graham (Sincere@april.biz) - hildegard.org
Ervin Howell (Shanna@melissa.tv) - anastasia.net
Clementine Bauch (Nathan@yesenia.net) - ramiro.info
Patricia Lebsack (Julianne.OConner@kory.org) - kale.biz
Chelsey Dietrich (Lucio_Hettinger@annie.ca) - demarco.info
Mrs. Dennis Schulist (Karley_Dach@jasper.info) - ola.org
Kurtis Weissnat (Telly.Hoeger@billy.biz) - elvis.io
Nicholas Runolfsdottir V (Sherwood@rosamond.me) - jacynthe.com
Glenna Reichert (Chaim_McDermott@dana.io) - conrad.com
Clementina DuBuque (Rey.Padberg@karina.biz) - ambrose.net


In [ ]:
import requests

api = requests.get('API')

json = api.json()

for moeda in json:
    nome = 'moed'
    preco = '$222'
    valor_mercado = '$555' 
    em_circulacao = '46465'
    percentual_dia = '1.56%'
    print('{0}: {1} ({2} / 24h)'.format(nome, preco, percentual_dia))
    print('Em circulação: {0}'.format(em_circulacao))
    print('Valor de mercado: {0}'.format(valor_mercado))
    print('-------------')

## Exercício 1 - Extraindo Dados do CoinMarketCap via API

Vamos usar o pacote requests para o acessar a API do CoinMarketCap.

As instruções da API estão disponíveis em: https://coinmarketcap.com/api/

Deveremos extrair os seguintes dados:
- Nome da moeda
- Preço atual **CONVERTIDO EM REAIS**
- O valor de mercado total
- O valor atual em circulação
- Percentual de mudança nas últimas 24 horas

Também deveremos limitar para que a API traga apenas 5 moedas para trabalharmos

Tempo: 10 minutos

---

In [15]:
import requests
import json

endpoint = 'https://api.coinmarketcap.com/v1/ticker/'

params = {'convert':'BRL',
          'limit':5}

api = requests.get(endpoint, params=params)

result = api.json()

msg = """{name}: {price_brl} ({percent_change_24h}% 24h) 
Valor atual em circulação: {available_supply}
Valor de mercado: {market_cap_brl}
--------------------------------------------------------""" 

for coin in result:
    name = coin['name']
    price_brl = coin['price_brl']
    percent_change_24h = coin['percent_change_24h']
    available_supply = coin['available_supply']
    market_cap_brl = coin['market_cap_brl']
    text = msg.format(name=name,
               price_brl=price_brl,
               percent_change_24h=percent_change_24h,
               available_supply=available_supply,
               market_cap_brl=market_cap_brl)
    print(text)

Bitcoin: 26339.957423 (-4.19% 24h) 
Valor atual em circulação: 16924337.0
Valor de mercado: 445786315993
--------------------------------------------------------
Ethereum: 1916.1884986 (-4.47% 24h) 
Valor atual em circulação: 98243763.0
Valor de mercado: 188253568239
--------------------------------------------------------
Ripple: 2.1586370901 (-5.36% 24h) 
Valor atual em circulação: 39091716516.0
Valor de mercado: 84384829187.0
--------------------------------------------------------
Bitcoin Cash: 3186.3842629 (1.71% 24h) 
Valor atual em circulação: 17022663.0
Valor de mercado: 54240743903.0
--------------------------------------------------------
Litecoin: 517.4292148 (-4.33% 24h) 
Valor atual em circulação: 55668706.0
Valor de mercado: 28804614895.0
--------------------------------------------------------


## Como utilizar APIs com autenticação

É muito comum encontrarmos APIs que necessitam de autenticação. Isto é, algum identificador que diz para ela quem está requisitando aquela informação.

Para realizarmos autenticação em APIs REST nós normalmente utilizamos um Header e um Token de acesso.

Porém, algumas APIs também utilizam parâmetros para realizar a autenticação.

Abaixo vamos ver como acessar a API da Battuta (https://battuta.medunes.net/) utilizando um token.

A Battuta é um serviço para dados geográficos.

*Aviso para leitores do futuro: se essa key não funcionar, é só entrar no site da battuta e gerar uma nova digitando o seu e-mail.*

In [17]:
import requests

payload = {'key': 'ce46369940c2632e9c160b0474bf49a3'}

api = requests.get('http://battuta.medunes.net/api/country/all/', params=payload)

json = api.json()

for i in range(10):
    country = json[i]
    print('{0} ({1})'.format(country['name'], country['code']))

Afghanistan (af)
Albania (al)
Algeria (dz)
Andorra (ad)
Angola (ao)
Anguilla (ai)
Antigua and Barbuda (ag)
Argentina (ar)
Armenia (am)
Aruba (aw)


## Testando o erro de não autorizado

In [12]:
import requests

payload = {'key': 'Key inválida.'}

api = requests.get('http://battuta.medunes.net/api/country/all/', params=payload)

print(''api.status_code)

json = api.json()

for i in range(10):
    country = json[i]
    print('{0} ({1})'.format(country['name'], country['code']))

403


KeyError: 0

## [4xx Client errors](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes)

This class of status code is intended for situations in which the error seems to have been caused by the client. Except when responding to a HEAD request, the server should include an entity containing an explanation of the error situation, and whether it is a temporary or permanent condition. These status codes are applicable to any request method. User agents should display any included entity to the user.

**400 Bad Request**

The server cannot or will not process the request due to an apparent client error (e.g., malformed request syntax, size too large, invalid request message framing, or deceptive request routing).

**401 Unauthorized (RFC 7235)**

Similar to 403 Forbidden, but specifically for use when authentication is required and has failed or has not yet been provided. The response must include a WWW-Authenticate header field containing a challenge applicable to the requested resource. See Basic access authentication and Digest access authentication. 401 semantically means "unauthenticated",[35] i.e. the user does not have the necessary credentials.
Note: Some sites issue HTTP 401 when an IP address is banned from the website (usually the website domain) and that specific address is refused permission to access a website.

**402 Payment Required**

Reserved for future use. The original intention was that this code might be used as part of some form of digital cash or micropayment scheme, as proposed for example by GNU Taler[36], but that has not yet happened, and this code is not usually used. Google Developers API uses this status if a particular developer has exceeded the daily limit on requests.

**403 Forbidden**

The request was valid, but the server is refusing action. The user might not have the necessary permissions for a resource, or may need an account of some sort.

**404 Not Found**

The requested resource could not be found but may be available in the future. Subsequent requests by the client are permissible.

**405 Method Not Allowed**

A request method is not supported for the requested resource; for example, a GET request on a form that requires data to be presented via POST, or a PUT request on a read-only resource.

**406 Not Acceptable**

The requested resource is capable of generating only content not acceptable according to the Accept headers sent in the request. See Content negotiation.

**407 Proxy Authentication Required (RFC 7235)**

The client must first authenticate itself with the proxy.

**408 Request Timeout**

The server timed out waiting for the request. According to HTTP specifications: "The client did not produce a request within the time that the server was prepared to wait. The client MAY repeat the request without modifications at any later time."

**409 Conflict**

Indicates that the request could not be processed because of conflict in the request, such as an edit conflict between multiple simultaneous updates.

Como você pode ver, utilizamos uma váriavel com todos os parâmetros que gostaríamos de passar.

O pacote requests colocou automaticamente o ponto de interrogação na URL para você!

---

Caso você se depare com uma API que utilize Headers, basta utilizar o exemplo abaixo como base.

---

In [1]:
import requests

h = {'my-custom-header': 'i am the one who knocks'}

r = requests.get('http://battuta.medunes.net/api/country/all/?key=ce46369940c2632e9c160b0474bf49a3', headers=h)

json = r.json()

for i in range(10):
    country = json[i]
    print('{0} ({1})'.format(country['name'], country['code']))

Afghanistan (af)
Albania (al)
Algeria (dz)
Andorra (ad)
Angola (ao)
Anguilla (ai)
Antigua and Barbuda (ag)
Argentina (ar)
Armenia (am)
Aruba (aw)


In [4]:
from pprint import pprint

In [5]:
pprint(r.headers)

{'Server': 'nginx', 'Date': 'Sat, 17 Mar 2018 13:56:42 GMT', 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Cache-Control': 'private, must-revalidate', 'pragma': 'no-cache', 'expires': '-1', 'Host-Header': '192fc2e7e50945beb8231a492d6a8024', 'X-Proxy-Cache': 'MISS'}


In [8]:
pprint(r.url)

'https://battuta.medunes.net/api/country/all/?key=ce46369940c2632e9c160b0474bf49a3'


---
## **Adendo: Autenticação usando OAuth1**

http://docs.python-requests.org/en/master/user/authentication/
https://github.com/requests/requests-oauthlib

```python
import requests
from requests_oauthlib import OAuth1

url = 'https://api.twitter.com/1.1/account/verify_credentials.json'
auth = OAuth1('YOUR_APP_KEY', 'YOUR_APP_SECRET',
              'USER_OAUTH_TOKEN', 'USER_OAUTH_TOKEN_SECRET')

requests.get(url, auth=auth)
```

**On github**

```python
from requests_oauthlib import OAuth1Session
twitter = OAuth1Session('client_key',
                        client_secret='client_secret',
                        resource_owner_key='resource_owner_key',
                        resource_owner_secret='resource_owner_secret')
url = 'https://api.twitter.com/1/account/settings.json'
r = twitter.get(url)
```

---

## **2. Escrita e leitura de arquivos em Python**


Em Python, assim como em outras linguagens de programação, podemos manipular arquivos que estão em nosso disco.

Para isso, utilizamos o `open` para nos trazer um **objeto de arquivo**.

O `open` precisa saber o nome do arquivo para conseguir ler seu conteúdo:

```python
open('nome_do_arquivo')
```

No entanto, existe um segundo argumento que podemos passar para dizer qual o **modo que queremos trabalhar com esse arquivo**.

Abaixo podemos ver exemplos de códigos com os modos mais comuns:

```python
open('nome_do_arquivo', 'r')  #Apenas leitura - padrão quando o modo é omitido

open('nome_do_arquivo', 'r+') #Leitura e escrita

open('nome_do_arquivo', 'w')  #Apenas escrita, apaga tudo que tem no arquivo antes de escrever novamente

open('nome_do_arquivo', 'a')  #"Append", adiciona o que você quiser ao final do arquivo, sem apagar nada
```

Também existe um argumento para passarmos em qual **encoding** gostaríamos de escrever.

```python
open('nome_do_arquivo', 'w', encoding='utf-8')
```

Você pode conferir todos os parâmetros do método `open` aqui: https://docs.python.org/3/library/functions.html#open

Todo o processo de escrita e leitura de arquivos deve ser feita com muito **cuidado**. Como são operações mais lentas, você pode acabar travando seu computador se fizer um loop errado enquanto escreve um arquivo.

Por isso, é muito importante se **lembrar de SEMPRE fechar o arquivo após o uso** dele no Python.

Vamos ver um exemplo:

In [7]:
f = open('arquivo_teste.txt', 'w')

f.write('Alô? Alguém me escuta?')

f.close()

## Porque o Context Manager (with) é necessário

Em Python, podemos pedir ajuda de um **Context Manager para automaticamente fechar o arquivo após o uso**.

Isso é a recomendação das **boas práticas** e evita que seu código tenha bugs.

Para utilizarmos essa funcionalidade, basta utilizar a palavra reservada **with**

Vamos ver um exemplo:

In [10]:
with open('arquivo_teste.txt', 'r') as f:
    print(f.readline())

print('\n\nSerá que o arquivo está fechado? {0}'.format(f.closed))

Alô? Alguém me escuta?


Será que o arquivo está fechado? True


## O que é um arquivo CSV

Arquivos CSV são apenas uma **coleção de dados separados por um caractere especial** (normalmente `,` ou `;`).

Eles são muito importantes para fazer **pequenos estudos de dados** ou movimentações de dados de um lugar para o outro.

Todos os softwares de planilha do mercado aceitam arquivos CSV.

Alguns cuidados que devemos seguir ao criar um arquivo CSV:

- Sempre terminar as linhas com `\n`
- Sempre utilizar o encoding como `utf-8`
- Utilizar caracteres que não influenciam no dado (ex.: vírgulas para separar dados númericos que também usam vírgulas)

Vamos ver um exemplo de como podemos criar um arquivo CSV em Python a partir do scraping da API do Jsonplaceholder.

In [13]:
import requests

api = requests.get('https://jsonplaceholder.typicode.com/users')

json = api.json()

with open('lista_de_usuarios.csv', 'w', encoding='utf-8') as f:
    for user in json:
        f.write('{0};{1};{2}\n'.format(user['name'], user['email'], user['website']))

        
# Vamos ler nosso arquivo
with open('lista_de_usuarios.csv', 'r') as r:
    for line in r:
        print(line)

Leanne Graham;Sincere@april.biz;hildegard.org

Ervin Howell;Shanna@melissa.tv;anastasia.net

Clementine Bauch;Nathan@yesenia.net;ramiro.info

Patricia Lebsack;Julianne.OConner@kory.org;kale.biz

Chelsey Dietrich;Lucio_Hettinger@annie.ca;demarco.info

Mrs. Dennis Schulist;Karley_Dach@jasper.info;ola.org

Kurtis Weissnat;Telly.Hoeger@billy.biz;elvis.io

Nicholas Runolfsdottir V;Sherwood@rosamond.me;jacynthe.com

Glenna Reichert;Chaim_McDermott@dana.io;conrad.com

Clementina DuBuque;Rey.Padberg@karina.biz;ambrose.net



## Exercício 2 - CSV com os dados do CoinMarketCap

Vamos usar o pacote requests para o acessar a API do CoinMarketCap e em seguida gravar o retorno em um CSV.

As instruções da API estão disponíveis em: https://coinmarketcap.com/api/

Deveremos extrair os seguintes dados:
- Nome da moeda
- Preço atual **CONVERTIDO EM REAIS**
- O valor de mercado total
- O valor atual em circulação
- Percentual de mudança nas últimas 24 horas

**Não precisaremos limitar a quantidade de moedas para esse exercício**

Tempo: 10 minutos

---

In [15]:
import requests

api = requests.get('API')

json = api.json()

#Abrir arquivo e escrever em padrão CSV

## Consumindo seu CSV

Agora que temos um CSV com muitos dados vamos importá-lo no Google Sheets e analisá-lo.

Entre em https://docs.google.com/spreadsheets/u/0/

Crie uma planilha nova

Vá no menu Arquivo e em seguida clique em Importar

![Como importar CSV no Google Sheets](images/importar_csv.png)

Em seguida será exibida uma tela para que você selecione a planilha que deseja importar.

Vá em Upload e selecione o arquivo em seu disco ou arraste para a janela.

![Como fazer upload de CSV no Google Sheets](images/upload_csv.png)

Depois será exibida uma tela pedindo para que você configure de acordo com o arquivo que você está importando.

![Como importar CSV no Google Sheets](images/escolher_padrao_csv.png)

Fique atento(a) para colocar o campo delimitador correto.

![Como importar CSV no Google Sheets](images/como_escolher_padrao_csv.png)

Pronto! Seus dados foram importados.

## Exercício 3 - CSV com os dados de uma fonte de sua escolha

**Atenção: este exercício será feito em dupla**

Escolham uma das APIs abaixo para depois guardar os dados em um CSV.

Fontes para API:
- https://jobs.github.com/api
- http://quotesondesign.com/wp-json/posts?filter[orderby]=rand&filter[posts_per_page]=25
- https://carbon-intensity.github.io/api-definitions/
- http://www.recipepuppy.com/about/api/
- https://app.swaggerhub.com/apis/Bandsintown/PublicAPI/3.0.0


Tempo: 20 minutos
    
---

## Parabéns

Terminamos por aqui. Não se esqueçam de checar o pós aula para mais dicas.

Qualquer dúvida meus contatos estão lá em cima!

![Acabou!](https://media.giphy.com/media/3o7qDEq2bMbcbPRQ2c/giphy.gif)